runtime/bundles/org.eclipse.e4.core.di/src/org/eclipse/e4/core/internal/di/InjectorImpl.java

```diff
@@ -148,38 +149,36 @@ private void internalInject(Object object, PrimaryObjectSupplier objectSupplier,
 
 	private void rememberInjectedObject(Object object, PrimaryObjectSupplier objectSupplier) {
 		synchronized (injectedObjects) {
-			List<WeakReference<?>> list = injectedObjects.computeIfAbsent(objectSupplier, k -> new ArrayList<>());
-			for (WeakReference<?> ref : list) {
-				if (object == ref.get())
-					return; // we already have it
+			Set<IdentityWeakReference<? extends Object>> map = injectedObjects.computeIfAbsent(objectSupplier,
+					k -> new LinkedHashSet<>());
+			IdentityWeakReference<?> ref = new IdentityWeakReference<>(object);
+			if (!map.contains(ref)) {
+				map.add(ref);
 			}
-			list.add(new WeakReference<>(object));
 		}
 	}
 
 	private boolean forgetInjectedObject(Object object, PrimaryObjectSupplier objectSupplier) {
 		synchronized (injectedObjects) {
-			List<WeakReference<?>> list = injectedObjects.get(objectSupplier);
+			Set<IdentityWeakReference<?>> list = injectedObjects.get(objectSupplier);
 			if (list != null) {
-				for (Iterator<WeakReference<?>> i = list.iterator(); i.hasNext();) {
-					WeakReference<?> ref = i.next();
-					if (object == ref.get()) {
-						i.remove();
-						return true;
-					}
+				IdentityWeakReference<?> ref = new IdentityWeakReference<>(object);
+				if (list.contains(ref)) {
```

Comment:
Here it will be more efficient to do a remove and use the boolean return from that.  Probably just:
```
return list.remove(ref);
```
Though as mentioned above list is a bad name now.  Probably ref can be inlined then too.